app/src/main/java/com/github/drawyourpath/bootcamp/authentication/FirebaseAuth.kt

```
@@ -0,0 +1,236 @@
+        oneTapClient = Identity.getSignInClient(activity)
+        signInRequest = BeginSignInRequest.builder()
+            .setPasswordRequestOptions(BeginSignInRequest.PasswordRequestOptions.builder()
+                .setSupported(true)
+                .build())
+            .setGoogleIdTokenRequestOptions(
+                BeginSignInRequest.GoogleIdTokenRequestOptions.builder()
+                    .setSupported(true)
+                    .setServerClientId(activity.getString(R.string.server_client_id))
+                    .setFilterByAuthorizedAccounts(true)
+                    .build())
+            .setAutoSelectEnabled(true)
+            .build()
+    }
+
+    private fun onGoogleSignInResult(activity: Activity, data: Intent?) {
+        try {
+            val account = GoogleSignIn.getSignedInAccountFromIntent(data).result
+            if (account != null && account.idToken != null) {
+                signInWithGoogleAuthToken(activity, account.idToken!!)
+            }
+        }
+        catch (e: java.lang.Exception) {
+            consumeCurrCallback(null, e)
+        }
+    }
+
+    private fun onOneTapSignInResult(activity: Activity, data: Intent?) {
+        try {
+            val credential = oneTapClient.getSignInCredentialFromIntent(data)
+            val idToken = credential.googleIdToken
+            when {
+                idToken != null -> {
+                    // Got an ID token from Google. Use it to authenticate
+                    // with your backend.
+                    signInWithGoogleAuthToken(activity, idToken)
+                }
+            }
+        } catch (e: ApiException) {
+            consumeCurrCallback(null, e)
+        }
+    }
+
+    override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent?) {
+        when (requestCode) {
+            REQ_GSI     -> onGoogleSignInResult(activity, data)
+            REQ_ONE_TAP -> onOneTapSignInResult(activity, data)
+        }
+    }
```

Maybe add error handling in case requestCode is not what we expect.

---

The method receives all the results of the activity so it might receive unrelated results with different request codes depending on the activity forwarding the results. Because of that, I don't think we should raise an error.